File: backend/gma.py
"""Gaussian moving average matching the Dual GMA Pine script in gmaxo.rtf."""

from __future__ import annotations

import numpy as np

SOURCE_PERIOD = 3
EMA_COL = "ema_3"
SMA_COL = "sma_3"


def sma(source: np.ndarray, period: int = SOURCE_PERIOD) -> np.ndarray:
    source = np.asarray(source, dtype=np.float64)
    n = source.size
    out = np.full(n, np.nan, dtype=np.float64)
    if period < 1:
        raise ValueError("period must be >= 1")
    if n < period:
        return out
    windows = np.lib.stride_tricks.sliding_window_view(source, period)
    out[period - 1 :] = windows.mean(axis=1)
    return out


def ema(source: np.ndarray, period: int = SOURCE_PERIOD) -> np.ndarray:
    """Pine-style EMA: seed with SMA, then alpha = 2 / (period + 1)."""
    source = np.asarray(source, dtype=np.float64)
    n = source.size
    out = np.full(n, np.nan, dtype=np.float64)
    if period < 1:
        raise ValueError("period must be >= 1")
    if n < period:
        return out
    alpha = 2.0 / (period + 1.0)
    out[period - 1] = source[:period].mean()
    for t in range(period, n):
        out[t] = alpha * source[t] + (1.0 - alpha) * out[t - 1]
    return out


def gaussian_weights(length: int, sigma_div: float) -> np.ndarray:
    if length < 1:
        raise ValueError("length must be >= 1")
    if sigma_div <= 0:
        raise ValueError("sigma_div must be > 0")
    i = np.arange(length, dtype=np.float64)
    x = i / (length / sigma_div)
    weights = np.exp(-0.5 * x * x)
    total = weights.sum()
    if total == 0:
        raise ValueError("gaussian weights summed to 0")
    return weights / total


def gaussian_ma(source: np.ndarray, length: int, sigma_div: float) -> np.ndarray:
    """Causal GMA: at bar t, weight source[t], source[t-1], ... source[t-length+1]."""
    source = np.asarray(source, dtype=np.float64)
    n = source.size
    out = np.full(n, np.nan, dtype=np.float64)
    if n < length:
        return out
    weights = gaussian_weights(length, sigma_div)
    windows = np.lib.stride_tricks.sliding_window_view(source, length)
    # windows[i] is source[i:i+length] (oldest -> newest). Pine uses newest first.
    out[length - 1 :] = windows[:, ::-1] @ weights
    return out


def dual_gma(
    close: np.ndarray,
    fast_length: int,
    fast_sigma: float,
    slow_length: int,
    slow_sigma: float,
    *,
    ema_source: np.ndarray | None = None,
    sma_source: np.ndarray | None = None,
) -> tuple[np.ndarray, np.ndarray]:
    """Fast GMA on EMA(close, 3); slow GMA on SMA(close, 3)."""
    fast_src = ema(close, SOURCE_PERIOD) if ema_source is None else ema_source
    slow_src = sma(close, SOURCE_PERIOD) if sma_source is None else sma_source
    fast = gaussian_ma(fast_src, fast_length, fast_sigma)
    slow = gaussian_ma(slow_src, slow_length, slow_sigma)
    return fast, slow


def detect_crosses(
    fast: np.ndarray, slow: np.ndarray
) -> tuple[np.ndarray, np.ndarray]:
    """Buy when fast crosses above slow; sell when fast crosses below slow."""
    n = min(len(fast), len(slow))
    buy = np.zeros(n, dtype=bool)
    sell = np.zeros(n, dtype=bool)
    if n < 2:
        return buy, sell
    f0, f1 = fast[1:n], fast[: n - 1]
    s0, s1 = slow[1:n], slow[: n - 1]
    valid = np.isfinite(f0) & np.isfinite(f1) & np.isfinite(s0) & np.isfinite(s1)
    buy[1:] = valid & (f1 <= s1) & (f0 > s0)
    sell[1:] = valid & (f1 >= s1) & (f0 < s0)
    return buy, sell

File: frontend/src/gma.ts
/** Dual GMA matching backend/gma.py and the optimizer worker: fast on EMA(close, 3), slow on SMA(close, 3). */

import { computeEma } from "./ema";
import type { GmaParams } from "./types";

const SOURCE_PERIOD = 3;

export function computeSma(source: Array<number | null>, period: number): Array<number | null> {
  const n = source.length;
  const out: Array<number | null> = Array(n).fill(null);
  if (period < 1 || n < period) return out;

  let start = 0;
  while (start < n && source[start] == null) start += 1;
  if (start + period > n) return out;

  let sum = 0;
  for (let i = 0; i < period; i++) {
    const value = source[start + i];
    if (value == null) return out;
    sum += value;
  }
  out[start + period - 1] = sum / period;
  for (let t = start + period; t < n; t++) {
    const incoming = source[t];
    const outgoing = source[t - period];
    const prev = out[t - 1];
    if (incoming == null || outgoing == null || prev == null) continue;
    out[t] = prev + (incoming - outgoing) / period;
  }
  return out;
}

/** Causal GMA: newest bar first, weights exp(-0.5 * (i / (length / sigma))^2). */
export function computeGaussianMa(
  source: Array<number | null>,
  length: number,
  sigma: number,
): Array<number | null> {
  const n = source.length;
  const out: Array<number | null> = Array(n).fill(null);
  if (length < 1 || sigma <= 0 || n < length) return out;

  const weights = new Float64Array(length);
  let total = 0;
  for (let i = 0; i < length; i++) {
    const x = i / (length / sigma);
    const weight = Math.exp(-0.5 * x * x);
    weights[i] = weight;
    total += weight;
  }
  if (total === 0) return out;
  for (let i = 0; i < length; i++) weights[i] /= total;

  for (let t = length - 1; t < n; t++) {
    let value = 0;
    let ok = true;
    for (let lag = 0; lag < length; lag++) {
      const sample = source[t - lag];
      if (sample == null || !Number.isFinite(sample)) {
        ok = false;
        break;
      }
      value += sample * weights[lag];
    }
    if (ok) out[t] = value;
  }
  return out;
}

export function computeDualGma(
  closes: number[],
  params: GmaParams,
): { fast: Array<number | null>; slow: Array<number | null> } {
  const n = closes.length;
  const empty = { fast: Array(n).fill(null) as Array<number | null>, slow: Array(n).fill(null) as Array<number | null> };
  if (
    n === 0 ||
    params.fastLength < 1 ||
    params.slowLength < 1 ||
    params.fastSigma <= 0 ||
    params.slowSigma <= 0
  ) {
    return empty;
  }
  const src: Array<number | null> = closes.map((value) => (Number.isFinite(value) ? value : null));
  const ema3 = computeEma(src, SOURCE_PERIOD);
  const sma3 = computeSma(src, SOURCE_PERIOD);
  return {
    fast: computeGaussianMa(ema3, params.fastLength, params.fastSigma),
    slow: computeGaussianMa(sma3, params.slowLength, params.slowSigma),
  };
}
